Add support for ipsec accelerator extern

usha1830 opened the pull request:
```
PNA extern for IPsec:
	// IPsec accelerator result for the current packet
	enum ipsec_status {
	    IPSEC_SUCCESS,
	    IPSEC_ERROR
	}
	 
	// IPsec accelerator extern definition
	extern ipsec_accelerator {
	                // IPsec constructor.
	                ipsec_accelerator();
	 
	                // Set the security association (SA) index for the current packet.
	                //
	                // When not invoked, the SA index to be used is undefined, leading to target dependent
	                // behavior.
	                void set_sa_index<T>(in T sa_index);
	 
	                // Set the offset to the IPv4/IPv6 header for the current packet.
	                //
	                // When not invoked, the default value set for the IPv4/IPv6 header offset is 0.
	                void set_ip_header_offset<T>(in T offset);
	 
	                // Enable the IPsec operation to be performed on the current packet.
	                //
	                // The type of the operation (i.e. encrypt/decrypt) and its associated parameters (such as
	                // the crypto cipher and/or authentication parameters, the tunnel/transport mode headers,
	                // etc) are completely specified by the SA that was set for the current packet.
	                //
	                // When enabled, this operation takes place once the deparser operation is completed.
	                void enable();
	 
	                // Disable any IPsec operation that might have been previously enabled for the current
	                // packet.
	                void disable();
	 
	                // Returns true when the current packet has been processed by the IPsec accelerator and
	                // reinjected back into the pipeline, and false otherwise.
	                bool from_ipsec(out ipsec_status status);
	}

```

1.  This extern is to be included in dpdk specific pna.p4
2. Pseudo Header for IPsec accelerator (platform_hdr_t)
   This a pseudo header inserted by the compiler with one field for security association index. This header needs to be instantiated in the main encapsulating header struct which is passed as argument to MainParser and MainControl

```
			T is a bit type 
			header platform_hdr_t {
			    T sa_id;
			}
			
			struct headers_t {
			    ... existing headers
				platform_hdr_t ipsec_hdr;
			}

```			
	
  Insert the pseudo header only if ipsec_accelerator is instantiated and enable is called.		
 3. IPSec needs 4 location registers. Size of these registers is 1 and these are set by the    control plane. 

	   ipsec_port_out_inbound
           ipsec_port_out_outbound
           ipsec_port_in_inbound
           ipsec_port_in_outbound


3. Method “set_sa_index”: sets the **ipsec_hdr.sa_id** field.
4. Method “enable”:
	a. Needs to validate the ipsec_instrisic header: **ipsec_hdr.setValid()**
        b. Needs to set the output port to the IPsec output port. This is done in the beginning of deparser, so that if there are multiple calls to enable and disable, the setting of output port for ipsec is done just once.

```
	1. if (ipsec_hdr.isValid()) 
		port_out = (direction == NET_TO_HOST) ? ipsec_inbound_port_out : ipsec_outbound_port_out ;
	2. //Deparser original start
	3. emit(ipsec_hdr)
	4. emit(rest of the headers)
```
							
6. Method “disable”:
	a. Needs to invalidate the ipsec_hdr: **ipsec_hdr.setInvalid()**
7. Method “from_ipsec”:
	a. Needs to test if the input port is equal to the IPsec accelerator port			
	b. Currently, for P4-DPDK, the status is always set to “success”.

```
status = SUCCESS
from_ipsec = (port_in == ipsec_port_in_inbound) && (port_in == ipsec_port_in_outbound)
```

usha1830 committed:
Add support for ipsec accelerator extern

kamleshbhalui commented on backends/dpdk/dpdkArch.cpp:
```diff
@@ -3035,4 +3035,169 @@ std::vector<std::pair<cstring, const IR::Type *>>
 cstring DpdkAddPseudoHeaderDecl::pseudoHeaderInstanceName;
 cstring DpdkAddPseudoHeaderDecl::pseudoHeaderTypeName;
 
+void InsertReqDeclForIPSec::uniqueNames() {
```
can we make it a utility? taking vector of cstring and refMap as arg?

usha1830 replied:
Done

kamleshbhalui commented on backends/dpdk/dpdkArch.cpp:
```diff
@@ -3035,4 +3035,169 @@ std::vector<std::pair<cstring, const IR::Type *>>
+}
+
+// create and add register declaration instance to program
+IR::Declaration_Instance *InsertReqDeclForIPSec::addRegDeclInstance(cstring instanceName) {
```
can we make it also generic and put in utility, all configurable part passed as param?

usha1830 replied:
You mean to add in DpdkUtils to be used by other passes as well?

kamleshbhalui commented on backends/dpdk/dpdkArch.cpp:
```diff
@@ -3035,4 +3035,169 @@ std::vector<std::pair<cstring, const IR::Type *>>
+    };
+
+    for (auto kv : structure->deparsers) {
+        if (kv.second->type->name == c->name && is_ipsec_used) {
```
can we use `is_ipsec_used` at top if false then return early in all visitor?

usha1830 replied:
Will do

kamleshbhalui commented on backends/dpdk/dpdkArch.h:
```diff
@@ -1320,5 +1320,91 @@ class CollectProgramStructure : public PassManager {
+            if (a->originalExternType->getName().name == "ipsec_accelerator") {
+                if (a->method->getName().name == "enable") {
+                    is_ipsec_used = true;
+                } else if (a->method->getName().name == "set_sa_index") {
```
why are doing this here, if goal of this pass is to check wether ipsec used or not?

usha1830 replied:
I will change the pass name to CollectIPSecInfo. This can be used to collect all information required like whether ipsed is used or not, what is the bitwidth for sa_id and ip_hdr_offset when we implement set_ip_header_offset.

psivanup reviewed (approved):
Looks fine to me.

Nice to have: Error category could have been used in error message.

usha1830 committed:
1) Add error types for all error methods
2) Error if ipsec_accelerator is used with PSA Arch
3) Fix Logical Not handling

usha1830 committed:
Refactor code, add utility functions

usha1830 committed:
Revert logical Not fix

usha1830 committed:
Update ipsec example program

usha1830 commented:
Closing as up-streamed
